planner/types.go

```diff
@@ -27,74 +27,48 @@ type PlanNode interface {
 	// Type of procedure represented by this node
 	Kind() ProcedureKind
 
-	// The types of the tables produced by this node
-	// Is it possible to know this at plan time?
-	// Type() []semantic.Type
-
 	// Helper methods for manipulating a plan
 	// These methods are used during planning
 	AddSuccessors(...PlanNode)
 	AddPredecessors(...PlanNode)
 	RemovePredecessor(PlanNode)
+	RemoveSuccessor(PlanNode)
 	ClearSuccessors()
 }
 
 type NodeID string
 
 // PlanSpec holds the result nodes of a query plan with associated metadata
 type PlanSpec struct {
-	results   map[PlanNode]struct{}
-	resources flux.ResourceManagement
-	now       time.Time
-}
-
-// CreatePlanSpec instantiates a new query plan with result nodes and metadata
-func CreatePlanSpec(results []PlanNode, resources flux.ResourceManagement, now time.Time) *PlanSpec {
-	plan := newPlanSpec(results)
-	plan.resources = resources
-	plan.now = now
-	return plan
-}
-
-func newPlanSpec(results []PlanNode) *PlanSpec {
-	r := make(map[PlanNode]struct{}, len(results))
-	for _, root := range results {
-		r[root] = struct{}{}
-	}
-	return &PlanSpec{results: r}
+	Roots     map[PlanNode]struct{}
+	Results   map[string]PlanNode
+	Resources flux.ResourceManagement
+	Now       time.Time
 }
 
-// Results returns the successor-less nodes of the query plan
-func (plan *PlanSpec) Results() []PlanNode {
-	roots := []PlanNode{}
-	for k := range plan.results {
-		roots = append(roots, k)
+// NewPlanSpec initializes a new query plan
+func NewPlanSpec() *PlanSpec {
+	return &PlanSpec{
+		Roots:   make(map[PlanNode]struct{}),
+		Results: make(map[string]PlanNode),
 	}
-	return roots
 }
 
-// Resources returns the resources used by the plan
-func (plan *PlanSpec) Resources() flux.ResourceManagement {
-	return plan.resources
-}
-
-// Now returns the plan's now time
-func (plan *PlanSpec) Now() time.Time {
-	return plan.now
-}
-
-// Replace replaces one of the result nodes of the query plan
-func (plan *PlanSpec) Replace(result, with PlanNode) {
-	delete(plan.results, result)
-	plan.results[with] = struct{}{}
+// Replace replaces one of the root nodes of the query plan
+func (plan *PlanSpec) Replace(root, with PlanNode) {
+	delete(plan.Roots, root)
```

Comment:
If this method is called during physical planning, then `Roots` and `Results` will become out of sync, right?

Reply:
True. I'm thinking ahead to when we have `LogicalPlanSpec` and `PhysicalPlanSpec`. Only the latter will have `Results`.